src/main/java/au/org/emii/ncdfgenerator/NcdfDefinitionXMLParser.java

```diff
@@ -1,276 +1,374 @@
+	class ParseDimension
 	{
-
-		if( isNodeName( node, "attribute"))
+		IDimension parse( Node node ) throws NcdfGeneratorException
 		{
-			Map< String, String> m = parseKeyVals( node );
-			String key = m.get("name");
-			String val = m.get("value");
-			return new SimpleImmutableEntry< String, String>( key, val );
+			if( !node.getNodeName().equals( "dimension" ))
+				throw new NcdfGeneratorException( "Not a dimension node" );
+
+			String name = "";
+			NamedNodeMap attrs = node.getAttributes();
+			for( int i = 0; i < attrs.getLength(); ++i )
+			{
+				Node child = attrs.item( i) ;
+				if( child.getNodeName().equals( "name" ))
+					name = Helper.nodeVal( child);
+		        else
+					throw new NcdfGeneratorException( "Unrecognized tag" );
+			}
+
+			return new DimensionImpl( name );
 		}
-		return null;
 	}
 
 
-	private Map<String, String> parseAttributes( Node node )
+	class ParseDimensions
 	{
-		if( isNodeName( node, "attributes"))
+		List< IDimension> parse( Node node ) throws NcdfGeneratorException
 		{
-			Map<String, String> m = new HashMap<String, String> ();
-			for( Node child : new NodeWrapper(node) ) {
-				SimpleImmutableEntry< String, String> pair = parseAttribute( child);
-				if( pair != null)
-					m.put( pair.getKey(), pair.getValue());
+			if( !node.getNodeName().equals( "dimensions" ))
+				throw new NcdfGeneratorException( "Not a dimensions node" );
+
+			List< IDimension> dimensions = new ArrayList< IDimension>();
+			for( Node child : new NodeWrapper(node) )
+			{
+				if( Helper.isElementNode( child)) {
+					String tag = child.getNodeName();
+					if( tag.equals( "dimension" ))
+						dimensions.add( new ParseDimension().parse( child ) );
+					else
+						throw new NcdfGeneratorException( "Unrecognized tag" );
+				}
 			}
-			return m;
+			return dimensions;
 		}
-		return null;
 	}
 
 
-	private IDimension parseDimensionRef ( Node node, Map< String, IDimension> dimensionsContext )
+	class ParseVariableDimension
 	{
-		if( isNodeName( node, "dimension")) {
-			Map< String, String> m = parseKeyVals( node );
-			return dimensionsContext.get( m.get( "name" ));
+		IDimension parse( Context context, Node node ) throws NcdfGeneratorException
+		{
+			if( !node.getNodeName().equals( "dimension" ))
+				throw new NcdfGeneratorException( "Not a dimension" );
+
+			NamedNodeMap attrs = node.getAttributes();
+			for( int i = 0; i < attrs.getLength(); ++i )
+			{
+				Node child = attrs.item( i) ;
+				if( child.getNodeName().equals( "name" ))
+					return context.getDimensionByName( Helper.nodeVal( child) );
+				else
+					throw new NcdfGeneratorException( "Unrecognized tag" );
+			}
+			return null;
 		}
-		return null;
 	}
 
 
-	private Map< String, IDimension> parseDimensionsRef( Node node, Map< String, IDimension> dimensionsContext )
+	class ParseVariableDimensions
 	{
-		if( isNodeName( node, "dimensions"))
+		List< IDimension> parse( Context context, Node node  ) throws NcdfGeneratorException
```

Comment:
For future reference - its not normal to prefix Interfaces with I in java.  You won't find any interfaces in java.lang etc  starting with I.  Take java.util.List for example, with concrete implementations java.util.ArrayList, java.util.LinkedList etc.